### ISSUE TITLE
Last upgrade yesterday and again today -> 100% CPU Emacs process

### ISSUE THREAD

**Issue by liar666** [doomemacs/doomemacs#7952]
### I confirm that...

- [X] I have searched the [issue tracker](https://github.com/doomemacs/doomemacs/issues), [documentation](https://docs.doomemacs.org), [FAQ](https://docs.doomemacs.org/-/faq), [Discourse](https://discourse.doomemacs.org), and [Google](https://google.com), in case this issue has already been reported/resolved.

- [X] I have read "[How to Debug Issues](https://doomemacs.org/d/how2debug)", and will use it to provide as much information about this issue as possible.

- [X] The issue can be reproduced on the **latest** available commit of Doom.

- [X] The issue can be reproduced on a stable release of Emacs, such as 27, 28, or 29. *(Unstable versions end in .50, .60, or .9x)*

### Expected behavior

`doom upgrade`ing should be flawless

### Current behavior

Running Emacs after upgrade yesterday and again today leads to 100% CPU for the Emacs process.

Running: `\emacs --debug-init`, I get a buffer that pops up with this message every second:
```
 ■  Warning (comp): code-review.el:91:2: Warning: custom-declare-variable `code-review-auth-login-marker' docstring has wrong usage of unescaped single quotes (use \= or different quoting)
 ■  Warning (comp): code-review.el:91:2: Warning: custom-declare-variable `code-review-auth-login-marker' docstring has wrong usage of unescaped single quotes (use \= or different quoting)
```

### Steps to reproduce

1. `doom upgrade`
2. (eventually `doom sync` `doom gc`)
3. \emacs --debug-init

### System Information

https://pastebin.com/xaB2dZKj

**Comment by liar666** [comment]
Update: Seems that waiting 10+ min solves the problem...
Not sure of what (Doom)Emacs is doing...
Not sure the error messages are related to the 100% CPU burst.

**Comment by k11m1** [comment]
I suspect this was the just in time native compilation and not a bug.

**Comment by hlissner** [comment]
This is, indeed, native-compilation. It is configured to natively compile elisp packages in the background, as they are loaded (assuming they haven't already been natively compiled). After a `doom upgrade` or `doom sync`, you likely have some freshly updated/installed packages that haven't been compiled yet, which is why it seems to kick in during/after `doom sync` and `doom upgrade`s.

Here are some examples for tweaking the behavior/resource intensity of native-comp:

```elisp
;;; in $DOOMDIR/config.el

;; This would disable native-compilation entirely.
(setq native-comp-deferred-compilation nil)

;; Or change how many CPUs/threads it should use. Doom defaults 
;; this to 1/4th of your cores. The Emacs default is to use 1/2 
;; your cores. This is how you set it to use 1/2 your cores:
(setq native-comp-async-jobs-number (max 1 (/ (num-processors) 2)))

;; How many optimizations native-comp should employ.
;;   -1 = no native-compilation at all
;;    0 = no optimizations at all
;;    1 = light optimizations
;;    2 = all (but only safe) optimizations  (the default)
;;    3 = all (including unsafe) optimizations
(setq native-comp-speed 1)

;; Or tell native-comp to ignore certain, problematic files:
(add-to-list 'native-comp-jit-compilation-deny-list "/some-file\\.e\\'")
```

**Comment by avdv** [comment]
I came here because I encountered the same problem but it would not go away... 

After upgrading my system and doom emacs, there was an `emacs --fg-daemon` process (started as a systemd user service) running at 100% CPU for hours. I could see that it would start child processes that would vanish very fast. There was also another emacs process (with `--daemon` flag) with a few child processes running. As soon as I killed the latter emacs process, the first one immediately started to behave differently, it would barely use CPU time, child processes started kept running for a few seconds and after a few minutes everything was done and my system was idle.

TLDR; make sure there is no other emacs daemon running in order for the native-compilation to make progress.

Maybe others find this useful.

**Comment by hlissner** [comment]
> there was an emacs --fg-daemon process (started as a systemd user service) running at 100% CPU for hours. I could see that it would start child processes that would vanish very fast

@avdv That doesn't seem related to this issue. native-comp, by default, won't use 100% of your CPU. Only 100% of 1/4 of your cores. It's concerning that something is spawning multiple `--daemon` processes. Neither Doom nor native-comp does this, so it sounds more like misconfiguration of the systemd service.
